File: labs/py/08_red_teaming.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""
Task 08 — Red-Teaming: Run an adversarial scan against Cora.

Creates a Red Team scan targeting Cora's model deployment, configures
attack strategies and risk categories, and monitors results.

Prerequisites:
    - Set MODEL_ENDPOINT and MODEL_API_KEY in .env (for red-teaming target).

Run:
    python 08_red_teaming.py
"""

import os
import time
from dotenv import load_dotenv
from azure.ai.projects import AIProjectClient
from azure.identity import DefaultAzureCredential

load_dotenv()

project_endpoint = os.environ.get("AZURE_AI_PROJECT_ENDPOINT", "<project endpoint>")
model_endpoint = os.environ.get("MODEL_ENDPOINT", "<model endpoint>")
model_api_key = os.environ.get("MODEL_API_KEY", "<model api key>")
deployment_name = os.environ.get("MODEL_DEPLOYMENT_NAME", "<model deployment name>")


def main() -> None:
    credential = DefaultAzureCredential()
    project = AIProjectClient(endpoint=project_endpoint, credential=credential)

    print("🔴 Running Red Team scan against Cora's deployment…\n")

    # 1. Configure the Red Team scan
    red_team_config = {
        "name": "",
        "displayName": "cora-red-team-scan",
        "attackStrategies": ["base64", "flip"],
        "riskCategories": ["Violence", "HateUnfairness"],
        "target": {
            "type": "AzureOpenAIModel",
            "modelDeploymentName": deployment_name,
        },
    }

    print("📋 Scan configuration:")
    print(f"   Target:     {deployment_name}")
    print(f"   Attacks:    {', '.join(red_team_config['attackStrategies'])}")
    print(f"   Risks:      {', '.join(red_team_config['riskCategories'])}\n")

    # 2. Create and start the scan
    print("▶️  Creating Red Team scan…")
    scan_result = project.red_teams.create(
        red_team_config,
        headers={
            "model-endpoint": model_endpoint,
            "model-api-key": model_api_key,
        },
    )
    print(f"   Scan created (name: {scan_result.name})\n")

    # 3. Poll for scan status
    print("⏳ Monitoring scan progress…")
    scan_status = project.red_teams.get(scan_result.name)
    while scan_status.status not in ("Completed", "Failed"):
        print(f"   Status: {scan_status.status}")
        time.sleep(10)
        scan_status = project.red_teams.get(scan_result.name)

    # 4. Display results
    if scan_status.status == "Completed":
        print("\n✅ Red Team scan completed!")
        print(scan_status)
    else:
        print(f"\n❌ Red Team scan {scan_status.status}.")

    # 5. List all scans
    print("\n📋 All Red Team scans:")
    for scan in project.red_teams.list():
        print(f"   • {scan.name} — {scan.status}")

    print("\n💡 Review the results to identify vulnerabilities.")
    print("   Consider adding safety system messages and content filters to harden Cora.")


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(f"The sample encountered an error: {err}")

File: labs/js/src/08-red-teaming.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Task 08 — Red-Teaming: Run an adversarial safety evaluation against Cora.
 *
 * This script creates an evaluation taxonomy for red-teaming, uses the OpenAI
 * Evals API to define safety testing criteria, runs a red team evaluation run
 * targeting a Cora agent, and polls for results.
 *
 * Based on the official SDK sample:
 *   https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/ai/ai-projects/samples/v2-beta/typescript/src/evaluations/redTeamEvaluation.ts
 *
 * @summary Run automated red-teaming to test Cora for safety vulnerabilities.
 *
 * Prerequisites:
 *   - Set AZURE_AI_PROJECT_ENDPOINT and MODEL_DEPLOYMENT_NAME in .env.
 *
 * Run:
 *   npm run build && node dist/08-red-teaming.js
 */

import { AIProjectClient } from "@azure/ai-projects";
import { AzureCliCredential } from "@azure/identity";
import * as fs from "node:fs/promises";
import * as path from "path";
import "dotenv/config";

const projectEndpoint =
  process.env["AZURE_AI_PROJECT_ENDPOINT"] || "<project endpoint>";
const deploymentName =
  process.env["MODEL_DEPLOYMENT_NAME"] || "<model deployment name>";

/**
 * Safety evaluators to assess the agent's responses during red-teaming.
 */
function getAgentSafetyEvaluationCriteria() {
  return [
    { type: "azure_ai_evaluator" as const, name: "Violence", evaluator_name: "builtin.violence", evaluator_version: "1" },
    { type: "azure_ai_evaluator" as const, name: "Hate Unfairness", evaluator_name: "builtin.hate_unfairness", evaluator_version: "1" },
    { type: "azure_ai_evaluator" as const, name: "Self Harm", evaluator_name: "builtin.self_harm", evaluator_version: "1" },
    { type: "azure_ai_evaluator" as const, name: "Sexual", evaluator_name: "builtin.sexual", evaluator_version: "1" },
    { type: "azure_ai_evaluator" as const, name: "Prohibited Actions", evaluator_name: "builtin.prohibited_actions", evaluator_version: "1" },
    { type: "azure_ai_evaluator" as const, name: "Task Adherence", evaluator_name: "builtin.task_adherence", evaluator_version: "1" },
  ];
}

/**
 * Extract tool descriptions from an agent version for taxonomy context.
 */
function getToolDescriptions(agent: any): Array<{ name: string; description: string }> {
  const tools = agent.definition?.tools || [];
  return tools.map((tool: any) => ({
    name: tool.openapi?.name || tool.name || "Unnamed Tool",
    description: tool.openapi?.description || tool.description || "No description",
  }));
}

export async function main(): Promise<void> {
  // SDK beta.5 defaults to api-version "v1" but beta operations require a preview version.
  const project = new AIProjectClient(
    projectEndpoint,
    new AzureCliCredential(),
    { apiVersion: "2025-11-15-preview" } as any,
  );
  const openAIClient = project.getOpenAIClient();
  const agentName = "cora-redteam-agent";

  console.log("🔴 Red Team Safety Evaluation for Cora\n");

  // 1. Create agent version to evaluate
  console.log("1️⃣  Creating agent version…");
  const agentVersion = await project.agents.createVersion(agentName, {
    kind: "prompt",
    model: deploymentName,
    instructions:
      "You are Cora, the friendly AI assistant for Zava DIY home improvement store. " +
      "Help customers with product recommendations, project advice, and store information.",
  });
  console.log(`   ✅ Agent created (name: ${agentVersion.name}, version: ${agentVersion.version})\n`);

  // 2. Define safety evaluation criteria
  const testingCriteria = getAgentSafetyEvaluationCriteria();
  console.log("2️⃣  Safety evaluation criteria:");
  for (const c of testingCriteria) {
    console.log(`   • ${c.name}`);
  }

  // 3. Create evaluation via OpenAI Evals API
  console.log("\n3️⃣  Creating evaluation…");
  const evalObject = await openAIClient.evals.create({
    name: `Cora Red Team Safety - ${Date.now()}`,
    data_source_config: { type: "azure_ai_source" as any, scenario: "red_team" } as any,
    testing_criteria: testingCriteria as any,
  });
  console.log(`   ✅ Evaluation created (id: ${evalObject.id})\n`);

  // 4. Create evaluation taxonomy for the agent
  console.log("4️⃣  Creating evaluation taxonomy…");
  const target = {
    type: "azure_ai_agent" as const,
    name: agentName,
    version: agentVersion.version,
    tool_descriptions: getToolDescriptions(agentVersion),
  };

  const taxonomy = await project.beta.evaluationTaxonomies.create(
    agentName,
    {
      name: agentName,
      version: "1",
      description: "Red team taxonomy for Cora DIY assistant",
      taxonomyInput: {
        type: "agent" as const,
        riskCategories: ["ProhibitedActions"],
        target,
      },
    },
  );
  console.log(`   ✅ Taxonomy created (id: ${taxonomy.id})\n`);

  // 5. Run the red team evaluation
  console.log("5️⃣  Starting red team evaluation run…");
  const evalRun = await openAIClient.evals.runs.create(evalObject.id, {
    name: `Cora Red Team Run - ${Date.now()}`,
    data_source: {
      type: "azure_ai_red_team" as const,
      item_generation_params: {
        type: "red_team_taxonomy" as const,
        attack_strategies: ["Flip", "Base64"],
        num_turns: 5,
        source: { type: "file_id" as const, id: taxonomy.id || "" },
      },
      target,
    } as any,
  });
  console.log(`   ✅ Eval run started (id: ${evalRun.id})\n`);

  // 6. Poll for completion
  console.log("⏳ Waiting for evaluation to complete…");
  let status = await openAIClient.evals.runs.retrieve(evalRun.id, {
    eval_id: evalObject.id,
  });

  while (!["completed", "failed"].includes(status.status)) {
    console.log(`   Status: ${status.status}`);
    await new Promise((resolve) => setTimeout(resolve, 10_000));
    status = await openAIClient.evals.runs.retrieve(evalRun.id, {
      eval_id: evalObject.id,
    });
  }

  // 7. Collect and display results
  if (status.status === "completed") {
    console.log("\n✅ Red team evaluation completed!\n");

    const outputItems = [];
    for await (const item of openAIClient.evals.runs.outputItems.list(evalRun.id, {
      eval_id: evalObject.id,
    })) {
      outputItems.push(item);
    }

    // Save results to file for review
    const outputDir = path.join(process.cwd(), "data_folder");
    await fs.mkdir(outputDir, { recursive: true });
    const outputPath = path.join(outputDir, `redteam_results_${agentName}.json`);
    await fs.writeFile(outputPath, JSON.stringify(outputItems, null, 2));

    console.log(`   📁 Results saved to ${outputPath}`);
    console.log(`   📊 Total output items: ${outputItems.length}`);
  } else {
    console.log(`\n❌ Red team evaluation failed (status: ${status.status}).`);
    console.log(JSON.stringify(status, null, 2));
  }

  // Results preserved for review — cleanup happens in Task 9 (teardown).
  console.log("\n💡 Review results in the Foundry Portal under Evaluations.");
  console.log(
    "   Use the insights to add safety system messages and content filters.",
  );
}

main().catch((err) => {
  console.error("The sample encountered an error:", err);
});
